packages/@aws-cdk/aws-appmesh/lib/tls-validation.ts

```diff
@@ -0,0 +1,218 @@
+
+  /**
+   * VirtualNode CFN configuration for listener's TLS Validation
+   *
+   * @default - no TLS Validation
+   */
+  readonly virtualNodeListenerTlsValidation?: CfnVirtualNode.ListenerTlsValidationContextProperty
+
+  /**
+   * VirtualGateway CFN configuration for client policy's TLS Validation
+   */
+  readonly virtualGatewayClientTlsValidation: CfnVirtualGateway.VirtualGatewayTlsValidationContextProperty;
+
+  /**
+   * VirtualGateway CFN configuration for listener's TLS Validation
+   *
+   * @default - no TLS Validation
+   */
+  readonly virtualGatewayListenerTlsValidation?: CfnVirtualGateway.VirtualGatewayListenerTlsValidationContextProperty;
+}
+
+/**
+ * Properties used to define TLS Validation.
+ */
+interface TLSValidationCommonOptions {
+  /**
+   * The value used to determine a SANs match and specified values must match exactly.
+   *
+   * @default - None
+   */
+  readonly subjectAlternativeNamesMatcherExact?: string[];
+}
+
+/**
+ * ACM Trust Properties
+ */
+export interface AcmTrustOptions extends TLSValidationCommonOptions {
+  /**
+   * Contains information for your private certificate authority
+   */
+  readonly certificateAuthorities: acmpca.ICertificateAuthority[];
+}
+
+/**
+ * File Trust Properties
+ */
+export interface FileTrustOptions extends TLSValidationCommonOptions {
+  /**
+   * Path to the Certificate Chain file on the file system where the Envoy is deployed.
+   */
+  readonly certificateChain: string;
+}
+
+/**
+ * SDS Trust Properties
+ */
+export interface SdsTrustOptions extends TLSValidationCommonOptions {
+  /**
+   * Path to the Certificate Chain file on the file system where the Envoy is deployed.
+   */
+  readonly secretName: string;
+}
+
+/**
+ * Defines the TLS validation context trust.
+ */
+export abstract class TlsValidation {
+  /**
+   * Tells envoy where to fetch the validation context from
+   */
+  public static fileTrust(props: FileTrustOptions): TlsValidation {
+    return new TlsValidationImpl(CertificateType.FILE, undefined,
+      props.certificateChain, undefined, props.subjectAlternativeNamesMatcherExact);
+  }
+
+  /**
+   * TLS validation context trust for ACM Private Certificate Authority (CA).
+   */
+  public static acmTrust(props: AcmTrustOptions): TlsValidation {
+    return new TlsValidationImpl(CertificateType.ACMPCA, props.certificateAuthorities,
+      undefined, undefined, props.subjectAlternativeNamesMatcherExact);
+  }
+
+  /**
+   * TLS validation context trust for Secret Discovery Service.
+   */
+  public static sdsTrust(props: SdsTrustOptions): TlsValidation {
+    return new TlsValidationImpl(CertificateType.SDS, undefined,
+      undefined, props.secretName, props.subjectAlternativeNamesMatcherExact);
+  }
+
+  /**
+   * Returns Trust context based on trust type.
+   */
+  public abstract bind(scope: Construct): TlsValidationConfig;
+}
+
+class TlsValidationImpl extends TlsValidation {
```

Comment:
So I remember you brought this up in our 1:1 that this was potentially an issue, i.e. `TlsCertificate` has options specific to the listener, like the `mode` field.

I wonder if we should actually model the `Certificate` and `Trust` as the union class and then have an interface for the validation context/listener specific differences. For example:

```typescript
interface ListenerTls {
  readonly certificate: Certificate;
  readonly tlsMode: TlsMode;
  readonly tlsValidation: TlsValidation; // this being the union class
}
```

and the client policy would look like:

```typescript
interface TlsValidationContext {
  readonly tlsValidation: TlsValidation; // contains both SAN matchers and trust bundle location
  readonly ports?: number[];
  readonly enforce?: bool;
  readonly certificate?: Certificate;
}
```


also, quick side note, for some reason `enforce` is not actually included in the current validation context. We should allow that to be configured..

Reply:
This is great point!
I agree creating the union type for `trust` property instead of `TlsValidation` property is more accurately representing the structure.